File: src/server/protocol.py
import socket
import struct
from enum import Enum
import uuid


class RequestCode(Enum):
    register = 1100
    users_list = 1101
    get_pub_key = 1102
    send_message = 1103
    get_messages = 1104


class Request:
    client_id_size = 16  # the size in bytes of client_id
    _chunk_size = 256

    def __init__(self, client_id, version, code, payload_size=0, payload=""):
        self.client_id = uuid.UUID(bytes_le=client_id)
        self.version = version
        self.code = RequestCode(code)
        self.payload_size = payload_size
        self.payload = payload

    def __str__(self):
        return f"client_id: {self.client_id}, version: {self.version}, " \
               f"code: {self.code.value} payload: {self.payload}"

    @staticmethod
    def process_request(conn):
        # read client id, Version, Code and Payload size
        pack_format = f"<{Request.client_id_size}sBHI"
        header_bytes = conn.recv(struct.calcsize(pack_format))
        data = struct.unpack(pack_format, header_bytes)
        i = struct.calcsize(pack_format)
        payload_size = data[-1]

        # if there is a payload
        if payload_size:
            payload_bytes = bytes()
            for i in range(0, payload_size, Request._chunk_size):
                read_size = min(Request._chunk_size, payload_size - i)
                payload_bytes += conn.recv(read_size)
            pack_format = f"<{payload_size}s"
            data += struct.unpack(pack_format, payload_bytes)
        req = Request(*data)
        return req


class ResponseCode(Enum):
    register = 2100
    users_list = 2101
    get_pub_key = 2102
    send_message = 2103
    get_messages = 2104
    error = 9000


class Response:
    def __init__(self, version, code=-1, payload=""):
        self.version = version
        try:
            self.code = ResponseCode(code)
        except ValueError:
            self.code = None
        # self.payload_size = payload_size
        self.payload_size = len(payload)
        self.payload = payload

    def __str__(self):
        return f"version: {self.version}, code: {self.code.value}, payload: {self.payload}"

    def send(self, conn):
        b = struct.pack("<BHI", self.version, self.code.value, self.payload_size)
        if self.payload_size:
            b += self.payload
        conn.sendall(b)

File: src/client/Protocol.cpp
#include "protocol.h"

Request::Request(UUID client_id, uint8_t version, requestCode _code, uint32_t _payload_size, char* _payload)
{
    setClientId(client_id);
    setVersion(version);
    setCode(_code);
    setPayloadSize(_payload_size);
    setPayload(_payload);
}

char* Request::getRequestBytes()
{
    char* request_data = new char[HEADER_SIZE + _payload_size];
    _bytes_amount = 0;

    if (request_data)
    {
        // copy the vars bytes
        std::memcpy(&request_data[_bytes_amount], &_client_id, sizeof(_client_id));
        _bytes_amount += sizeof(_client_id);
        std::memcpy(&request_data[_bytes_amount], &_version, sizeof(_version));
        _bytes_amount += sizeof(_version);
        std::memcpy(&request_data[_bytes_amount], &_code, sizeof(_code));
        _bytes_amount += sizeof(_code);
        std::memcpy(&request_data[_bytes_amount], &_payload_size, sizeof(_payload_size));
        _bytes_amount += sizeof(_payload_size);
        // validate that there is a payload
        if (_payload_size != 0)
        {
            std::memcpy(&request_data[_bytes_amount], _payload, _payload_size);
            _bytes_amount += _payload_size;
        }
    }

    return request_data;
}

Response Request::sendRequset(std::string ip, int port)
{
    WSADATA wsaData;
    int iResult = WSAStartup(MAKEWORD(2, 2), &wsaData);
    if (iResult != 0) {
        std::string err_msg = "WSAStartup failed with error: " + iResult;
        throw std::exception(err_msg.c_str());
    }
    SOCKET ConnectSocket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP); // create socket
    if (ConnectSocket == INVALID_SOCKET) {
        std::string err_msg = "Error at socket(): " + WSAGetLastError();
        throw std::exception(err_msg.c_str());
    }
    struct sockaddr_in saServer;
    saServer.sin_family = AF_INET;
    inet_pton(AF_INET, ip.c_str(), &saServer.sin_addr.s_addr);
    saServer.sin_port = htons(port);
    iResult = connect(ConnectSocket, (SOCKADDR*)&saServer, sizeof(saServer)); // connect
    if (iResult == SOCKET_ERROR) {
        closesocket(ConnectSocket);
        ConnectSocket = INVALID_SOCKET;
        std::string err_msg = "failed to connect";
        throw std::exception(err_msg.c_str());
    }
    char* request_bytes = getRequestBytes(); // get the request's bytes
    iResult = send(ConnectSocket, request_bytes, _bytes_amount, 0); // send the request to the server
    delete request_bytes;
    if (iResult == SOCKET_ERROR) {
        closesocket(ConnectSocket);
        std::string err_msg = "send failed with error: " + WSAGetLastError();
        throw std::exception(err_msg.c_str());
    }
    iResult = shutdown(ConnectSocket, SD_SEND); // shutdown the connection for sending more data
    if (iResult == SOCKET_ERROR) {
        std::string err_msg = "shutdown failed with error: " + WSAGetLastError();
        closesocket(ConnectSocket);
        WSACleanup();
        throw std::exception(err_msg.c_str());
    }
    Response resp = Response::processResponse(ConnectSocket); // procces the response from the server
    closesocket(ConnectSocket);
    WSACleanup();
    return resp;
}

Response Response::processResponse(SOCKET ConnectSocket)
{
    char* header_bytes = new char[Response::HEADER_SIZE];
    int recived_bytes = recv(ConnectSocket, header_bytes, Response::HEADER_SIZE, 0);
    Response resp = Response::processResponseHeader(header_bytes); // get the header as Response objet
    resp._payload = new char[resp._payload_size](); // create the payload
    // read the data from the server in chunks
    for (uint32_t i = 0;i < resp._payload_size;i += CHUNK_SIZE)
    {
        int read_size = std::min<int>(CHUNK_SIZE, resp._payload_size - i);
        recived_bytes = recv(ConnectSocket, &resp._payload[i], read_size, 0);
    }
    return resp;
}

Response Response::processResponseHeader(char serverdata[HEADER_SIZE])
{
    Response(resp);
    resp._bytes_amount = 0;
    std::memcpy(&resp._version, &serverdata[resp._bytes_amount], sizeof(resp._version));
    resp._bytes_amount += sizeof(resp._version);
    std::memcpy(&resp._code, &serverdata[resp._bytes_amount], sizeof(resp._code));
    resp._bytes_amount += sizeof(resp._code);
    std::memcpy(&resp._payload_size, &serverdata[resp._bytes_amount], sizeof(resp._payload_size));
    resp._bytes_amount += sizeof(resp._payload_size);
    return resp;
}
